internal/jobexecutor/job_executor.go

```diff
@@ -171,6 +176,57 @@ func (e *JobExecutor) execute(ctx context.Context) (res *jobExecutorResult) {
 	metadataUpdates := make(map[string]any)
 	ctx = context.WithValue(ctx, ContextKeyMetadataUpdates, metadataUpdates)
 
+	// Watches for jobs that may have become stuck. i.e. They've run longer than
+	// their job timeout (plus a small margin) and don't appear to be responding
+	// to context cancellation (unfortunately, quite an easy error to make in
+	// Go).
+	//
+	// Currently we don't do anything if we notice a job is stuck. Knowing about
+	// stuck jobs is just used for informational purposes in the producer in
+	// generating periodic stats.
+	if e.ClientJobTimeout > 0 {
+		// We add a WithoutCancel here so that this inner goroutine becomes
+		// immune to all context cancellations _except_ the one where it's
+		// cancelled because we leave JobExecutor.execute.
+		ctx, cancel := context.WithCancel(context.WithoutCancel(ctx))
+		defer cancel()
```

Comment:
May be worth calling out the intentional shadowing here too or using an alternate `ctx` var name